### PR TITLE
feat: add `use_https_scheme` for Windows and Android

### LINKED ISSUE #11252: [bug]  [v2] IndexedDB path changed after upgrading to tauri 2
### Describe the bug

After upgrading from tauri 1 to tauri 2, the IndexedDB directory changed from `https_tauri.localhost_0.indexeddb.leveldb` to `http_tauri.localhost_0.indexeddb.leveldb`. Since my app stores all data in IndexedDB, this will cause "data losing" from user's point of view.

This happens on Windows, the full paths are:

- `C:\Users\[username]\AppData\Local\[appid]\EBWebView\Default\IndexedDB\https_tauri.localhost_0.indexeddb.leveldb`
- `C:\Users\[username]\AppData\Local\[appid]\EBWebView\Default\IndexedDB\http_tauri.localhost_0.indexeddb.leveldb`

### Reproduction

To verify that tauri 2 uses said IndexedDB directory `http_tauri.localhost_0.indexeddb.leveldb`, following these steps:

1. create a new app: `pnpm create tauri-app`
2. do anything that causes IndexedDB directory being created, for example, using `idb`:
  a. `pnpm i idb`
  b. open IndexedDB: `openDB("test").then((db) => console.log("db opened: ", db));`
3. build: `pnpm tauri build --debug`
4. run the app and check the directory

I've set up a repo at https://github.com/hut36/tauri2-indexeddb for you to test.

### Expected behavior

The directory `https_tauri.localhost_0.indexeddb.leveldb` is expected, instead of `http_tauri.localhost_0.indexeddb.leveldb`. See above for the full path.

### Full `tauri info` output

```text
[✔] Environment
    - OS: Windows 10.0.22631 x86_64 (X64)
    ✔ WebView2: 125.0.2535.92
    ✔ MSVC: Visual Studio 生成工具 2022
    ✔ rustc: 1.81.0 (eeb90cda1 2024-09-04)
    ✔ cargo: 1.81.0 (2dbb1af80 2024-08-20)
    ✔ rustup: 1.27.1 (54dd3d00f 2024-04-24)
    ✔ Rust toolchain: stable-x86_64-pc-windows-msvc (default)
    - node: 20.12.0
    - pnpm: 9.10.0
    - yarn: 1.22.22
    - npm: 9.6.3

[-] Packages
    - tauri 🦀: 2.0.1
    - tauri-build 🦀: 2.0.1
    - wry 🦀: 0.44.1
    - tao 🦀: 0.30.3
    - @tauri-apps/api : 2.0.1
    - @tauri-apps/cli : 2.0.1

[-] Plugins
    - tauri-plugin-shell 🦀: 2.0.1
    - @tauri-apps/plugin-shell : 2.0.0

[-] App
    - build-type: bundle
    - CSP: unset
    - frontendDist: ../build
    - devUrl: http://localhost:1420/
    - framework: Svelte
    - bundler: Vite
```

### Stack trace

_No response_

### Additional context

Related discussion: https://github.com/tauri-apps/tauri/discussions/11231. @FabianLars says this is related to `dangerousUseHttpScheme`.

### PR DESCRIPTION
**Pull request by amrbashir**
closes #11252

### CODE REVIEW

File: crates/tauri-runtime/src/webview.rs [hunk 1]
```diff
@@ -378,6 +381,21 @@ impl WebviewAttributes {
     self.browser_extensions_enabled = enabled;
     self
   }
+
+  /// Sets whether the custom protocols should use `https://<scheme>.localhost` instead of the default `http://<scheme>.localhost` on Windows and Android. Defaults to `false`.
+  ///
+  /// ## Note
+  ///
+  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and is therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```
Comment (FabianLars) [inline crates/tauri-runtime/src/webview.rs]:
```suggestion
  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```

File: crates/tauri-utils/src/config.rs [hunk 2]
```diff
@@ -1486,6 +1486,18 @@ pub struct WindowConfig {
   /// - **MacOS / Linux / iOS / Android** - Unsupported.
   #[serde(default)]
   pub browser_extensions_enabled: bool,
+
+  /// Sets whether the custom protocols should use `https://<scheme>.localhost` instead of the default `http://<scheme>.localhost` on Windows and Android. Defaults to `false`.
+  ///
+  /// ## Note
+  ///
+  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and is therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```
Comment (FabianLars) [inline crates/tauri-utils/src/config.rs]:
```suggestion
  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```
(regen schema afterwards)

File: crates/tauri/src/webview/mod.rs [hunk 3]
```diff
@@ -787,6 +787,21 @@ fn main() {
     self.webview_attributes.browser_extensions_enabled = enabled;
     self
   }
+
+  /// Sets whether the custom protocols should use `https://<scheme>.localhost` instead of the default `http://<scheme>.localhost` on Windows and Android. Defaults to `false`.
+  ///
+  /// ## Note
+  ///
+  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and is therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```
Comment (FabianLars) [inline crates/tauri/src/webview/mod.rs]:
```suggestion
  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```

File: crates/tauri/src/webview/webview_window.rs [hunk 4]
```diff
@@ -896,6 +896,21 @@ impl<'a, R: Runtime, M: Manager<R>> WebviewWindowBuilder<'a, R, M> {
     self.webview_builder = self.webview_builder.browser_extensions_enabled(enabled);
     self
   }
+
+  /// Sets whether the custom protocols should use `https://<scheme>.localhost` instead of the default `http://<scheme>.localhost` on Windows and Android. Defaults to `false`.
+  ///
+  /// ## Note
+  ///
+  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and is therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```
Comment (FabianLars) [inline crates/tauri/src/webview/webview_window.rs]:
```suggestion
  /// Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```

File: packages/api/src/webview.ts [hunk 5]
```diff
@@ -728,6 +728,21 @@ interface WebviewOptions {
    * - **Android / iOS**: Unsupported.
    */
   zoomHotkeysEnabled?: boolean
+
+  /**
+   * Sets whether the custom protocols should use `https://<scheme>.localhost` instead of the default `http://<scheme>.localhost` on Windows and Android. Defaults to `false`.
+   *
+   * #### Note
+   *
+   * Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and is therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```
Comment (FabianLars) [inline packages/api/src/webview.ts]:
```suggestion
   * Using a `https` scheme will NOT allow mixed content when trying to fetch `http` endpoints and therefore will not match the behavior of the `<scheme>://localhost` protocols used on macOS and Linux.
```

**Review by FabianLars — COMMENTED** [review]
no need to apply the suggestions, i know it's easier to do locally (esp since you have to regen the schema).

Also this part
```
Changing this value between releases will change the IndexedDB, cookies and localstorage location and your app will not be able to access them.
```
is a bit off, i have to assume that the last part will be interpreted as "no indexeddb access at all" by a few non native speakers.

What about something simple like this?
```
Changing this value between releases will change the IndexedDB, cookies and localstorage location and your app will not be able to access the old data.
```

**Review by FabianLars — DISMISSED** [review]

**Comment by FabianLars** [comment]
for the migration: what about the dangerousUseHttpsScheme config? Since it's marked dangerous i wouldn't mind ignoring it, just asking in case you forgot that config.

**Comment by amrbashir** [comment]
will add that as well

**Review by FabianLars — DISMISSED** [review]

**Review by lucasfernog — DISMISSED** [review]

**Review by lucasfernog — APPROVED** [review]

**Pull request merged**